File: scilab/modules/linear_algebra/src/fortran/intdsyev.f
      subroutine intdsyev(fname)

c     [D,V]=spec(A)

      include 'stack.h'
      logical getrhsvar,createvar
      logical checklhs,checkrhs
      integer vfinite

      character fname*(*)
      character JOBZ, UPLO
      double precision ZERO
      parameter ( ZERO=0.0D0 )
c     
      minrhs=1
      maxrhs=1
      minlhs=1
      maxlhs=2
c     
      if(.not.checkrhs(fname,minrhs,maxrhs)) return
      if(.not.checklhs(fname,minlhs,maxlhs)) return 

      if(.not.getrhsvar(1,'d', M, N, lA)) return
      if(M.ne.N) then
         err=1
         call error(20)
         return
      endif
      if(N.eq.0) then
         if(lhs.eq.1) then
            lhsvar(1) = 1
            return
         else if(lhs.eq.2) then
            if(.not.createvar(2,'d', N, N, lV)) return
            lhsvar(1) = 2
            lhsvar(2) = 1
            return
         endif
      endif
      if(vfinite(M*N,stk(lA)).eq.0) then
         err=1
         call error(264)
         return
      endif 
      if(lhs.eq.1) then
         if(.not.createvar(2,'d', N, 1, lD)) return              
      else
         if(.not.createvar(2,'d', N, N, lD)) return
      endif
      if(.not.createvar(3,'d', N, 1, lW)) return
      LWORKMIN = 3*N - 1
      LWORK=maxvol(4,'d')
      if(LWORK.le.LWORKMIN) then
         err=(LWORK-LWORKMIN)
         call error(17)
         return
      endif
      if(.not.createvar(4,'d',1,LWORK,lDWORK)) return
      if(lhs.eq.1) then
         JOBZ = 'N'
      else
         JOBZ = 'V'
      endif 
      UPLO = 'U'
      call DSYEV( JOBZ, UPLO, N, stk(lA), N, stk(lW),
     $     stk(lDWORK), LWORK, INFO )
c     SUBROUTINE DSYEV( JOBZ, UPLO, N, A, LDA, W, WORK, LWORK, INFO )
      if(info.ne.0) then
         call error(24)
         return
      endif

      if(lhs.eq.1) then
         do 10 i = 1, N
            stk(lD+i-1) = stk(lW+i-1)
 10      continue   
         lhsvar(1)=2  
      else 
         call DLASET( 'F', N, N, ZERO, ZERO, stk(lD), N ) 
c     SUBROUTINE DLASET( UPLO, M, N, ALPHA, BETA, A, LDA )

         do 20 i = 1, N
            ii = i+(i-1)*N
            stk(lD+ii-1) = stk(lW+i-1)
 20      continue   
         lhsvar(1)=1
         lhsvar(2)=2

      endif 
c     
      end

File: scilab/modules/linear_algebra/examples/intdsyev.c
/*
* Scilab ( http://www.scilab.org/ ) - This file is part of Scilab
* Copyright (C) INRIA
* Copyright (C) DIGITEO - 2010 - Allan CORNET
*
* This file is released under the 3-clause BSD license. See COPYING-BSD.
*/
/* -------------------------------------------------------------------------- */
/* WARNING:this gateway uses old scilab api and it does not work with Scilab 6*/
/* see help api_scilab for more information. */
/* -------------------------------------------------------------------------- */
#include "sciprint.h"
#include "Scierror.h"
#include "stack-c.h"
#include "machine.h"
#include "core_math.h"
/* -------------------------------------------------------------------------- */
/* SUBROUTINE DSYEV( JOBZ, UPLO, N, A, LDA, W, WORK, LWORK, INFO ) */
/* [w,U] = dsyev(A) */
/* w = dsyev(A) */
/* -------------------------------------------------------------------------- */
extern int C2F(dsyev)();
/* -------------------------------------------------------------------------- */
int intdsyev(char* fname)
{
    int M = 0, N = 0, lA = 0;
    int NLHS = 0;
    int un = 1;
    int WORK = 0, lWORK = 0, LWORKMIN = 0, LWORK = 0, INFO = 0;
    int LDA = 0, lw = 0;
    int A = 1, W = 2;

    static int minlhs = 1, minrhs = 1, maxlhs = 2, maxrhs = 1;
    CheckRhs(minrhs, maxrhs) ;
    CheckLhs(minlhs, maxlhs) ;

    /*--------------------A---------------------------*/
    GetRhsVar(A, MATRIX_OF_DOUBLE_DATATYPE, &M, &N, &lA);
    /*  To be done: Check A symmetric and Real*/
    NLHS = Lhs;
    switch ( NLHS )
    {

        case 2:
            /*  [w,U]=dsyev(A)     */
            CreateVar(W, MATRIX_OF_DOUBLE_DATATYPE, &N, &un, &lw);
            WORK = 3;
            LWORKMIN = Max(1, 3 * N - 1);
            LWORK = LWORKMIN;
            /* LWORK=C2F(maxvol)(&WORK, "d", 1L);   max memory currently available */
            if (LWORK < LWORKMIN)
            {
                sciprint("not enough memory (use stacksize)");
                SciError(9999);
            };
            CreateVar(3, MATRIX_OF_DOUBLE_DATATYPE, &LWORK, &un, &lWORK);
            LDA = Max(1, N);
            C2F(dsyev)("V", "L", &N, stk(lA), &LDA, stk(lw), stk(lWORK), &LWORK, &INFO);
            if (INFO != 0)
            {
                C2F(errorinfo)("dsyev ", &INFO, 5L);
            }
            LhsVar(1) = A;
            LhsVar(2) = W;
            return 0;
            break;

        case 1:
            /*   w=dsyev(A)    */
            CreateVar(W, MATRIX_OF_DOUBLE_DATATYPE, &N, &un, &lw);
            WORK = 3;
            LWORKMIN = Max(1, 3 * N - 1);
            LWORK = C2F(maxvol)(&WORK, "d", 1L); /* max memory currently available */
            if (LWORK < LWORKMIN)
            {
                sciprint("not enough memory (use stacksize)");
                SciError(9999);
            };
            CreateVar(3, MATRIX_OF_DOUBLE_DATATYPE, &LWORK, &un, &lWORK);
            LDA = Max(1, N);
            C2F(dsyev)("N", "L", &N, stk(lA), &LDA, stk(lw), stk(lWORK), &LWORK, &INFO);
            if (INFO != 0)
            {
                C2F(errorinfo)("dsyev ", &INFO, 5L);
            }
            LhsVar(1) = W;
            return 0;
            break;

        default:
            return 0;
    }
    return(0);
}
/* -------------------------------------------------------------------------- */
